Add serial monitor functionality.

zck opened the pull request:
This commit adds functionality for arduino's serial monitor, including sending input from the buffer to the arduino (by typing something and pressing enter). This helps with debugging or interactive applications.

If the serial monitor is on when an upload happens, this code handles stopping the serial monitor before the upload, and restarting it after -- but only restarting if the upload succeeds. It reuses the same buffer for the monitor, so the history is shown.

**New keybindings in the keymap:**
* `m`: start the monitor. This handles stopping the monitor if it's open.
* `M`: stop the monitor.

**Future changes:**
1. Because the keybindings are in the regular keymap, they don't work inside the monitor buffer. This will be a future change I would like to make, but I don't think it's necessary to merge this in.
2. Only allow the user to provide input on the last line. Right now, any line is editable, and when you press enter, the whole line will be sent, even if that line was previously output.

zck committed:
Add serial monitor functionality.

ahihi commented:
thank you! this is super useful and i have long been meaning to implement it myself, now i dont need to :) showing the reason for stopping the monitor is a nice touch.

just one thing: you are currently passing the port and baud rate arguments directly to the shell - these should be escaped with `shell-quote-argument` to avoid any nasty surprises.

zck committed:
pass args to monitor to shell-quote-argument.

zck committed:
When starting the monitor, say how to stop it.

zck committed:
Add open, close serial monitor to the readme keymap.

zck commented:
Awesome, thank you! I `shell-quote-argument`ed the port and baud rate -- thanks for that information!

I also added two other commits:

* When the `arduino-cli monitor` command starts, it prints out `Connecting to /dev/ttyACM0. Press CTRL-C to exit.`. But `C-c` doesn't work, so I put a message in the monitor buffer about how to stop the monitor. I thought about passing `--quiet` to the monitor command, but then the user wouldn't know it started up. I don't have a better solution.
* Added the open, close commands to the readme.

ahihi commented:
great!

> * When the `arduino-cli monitor` command starts, it prints out `Connecting to /dev/ttyACM0. Press CTRL-C to exit.`. But `C-c` doesn't work, so I put a message in the monitor buffer about how to stop the monitor. I thought about passing `--quiet` to the monitor command, but then the user wouldn't know it started up. I don't have a better solution.

i think this is fine, but it reminds me that we do have verbosity related options `arduino-cli-verbosity` and `arduino-cli-compile-only-verbosity`. it would make sense for the serial monitor to respect these too. this is easily achieved with `arduino-cli--add-flags`, see e.g. `arduino-cli--compile` for an example of how it is used.

zck committed:
Serial monitor connections can be verbose or quiet.

zck commented:
Ooh, I'd missed that part! I just pushed a change that uses the `arduino-cli--add-flags` method, and also skips printing the "starting" and "stopping" messages when `arduino-cli-verbosity` is `'quiet`.

zck committed:
Fix the serial monitor callback using the right hook.

ahihi commented:
nice, thank you! merging :)

zck commented:
Hooray! I'm so happy to be able to do arduino work in Emacs! Now I can compile, upload, and interact with it!

I just need to figure out some issues I'm having with arduino-mode sometimes loses text, but that's unrelated to this repo entirely!

Thanks so much for merging it!
